{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "___\n",
    "# Jovian analysis report\n",
    "___"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**This report should be generated automatically. However, if it somehow doesn't then please follow the steps below to view your results:**  \n",
    "<br> \n",
    "**- You can generate the report by pressing `Cell` in the toolbar and then selecting `Run All`. This can take a couple of minutes (depending on the size of your dataset). As long as the circle in the top right corner is not-empty (i.e. completely filled), the data is still being loaded.**  \n",
    "<br>\n",
    "**- If you want to show or hide the programming code, press the button with the eye in it on the toolbar above.**  \n",
    "<br>\n",
    "**- N.B. The sum total of reads in this report will not add up to the sum total number of reads that were supplied as input. This is because, 1) human reads are removed, 2) PCR-duplicate reads are removed.**  \n",
    "<br>\n",
    "You can view the graph displaying all Jovian's disparate processes by clicking [here](https://github.com/DennisSchmitz/Jovian/blob/assets/images/rulegraph_Jovian.png?raw=true)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "deleteable": false,
    "init_cell": true,
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "%%html\n",
    "<style>\n",
    "div.output_subarea {\n",
    "    padding-top: 0 !important;\n",
    "    padding-bottom: 0 !important;\n",
    "}\n",
    "\n",
    ".standardheader {\n",
    "    font-size: 30px !important;\n",
    "    padding: 0px;\n",
    "}\n",
    "\n",
    ".standardanchor {\n",
    "    font-size: 14px !important;\n",
    "    margin-left: 5px;\n",
    "}\n",
    ".standardtext {\n",
    "    font-family: sans-serif;\n",
    "    font-size: 14px !important;\n",
    "    margin-left: 5px;\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "deleteable": false,
    "init_cell": true,
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "######################################\n",
    "# Required packages for this script  #\n",
    "######################################\n",
    "import pandas as pd\n",
    "import qgrid\n",
    "import glob\n",
    "import os\n",
    "from IPython.display import display as dp\n",
    "from IPython.display import Markdown as md\n",
    "from IPython.display import IFrame as fr\n",
    "from IPython.display import HTML as ht\n",
    "import plotly as pl\n",
    "import plotly.graph_objects as go\n",
    "import plotly.express as px\n",
    "\n",
    "\n",
    "### Do Bash things\n",
    "!bash bin/scripts/set_symlink.sh\n",
    "!bash bin/includes/Start_nginx start\n",
    "\n",
    "### Standard Qgrid options\n",
    "grid_options = {\n",
    "    'fullWidthRows': True,\n",
    "    'syncColumnCellResize': True,\n",
    "    'forceFitColumns': False,\n",
    "    'defaultColumnWidth': 100,\n",
    "    'rowHeight': 23,\n",
    "    'enableColumnReorder': True,\n",
    "    'enableTextSelectionOnCells': True,\n",
    "    'editable': True,\n",
    "    'autoEdit': False,\n",
    "    'explicitInitialization': True,\n",
    "    'maxVisibleRows': 20,\n",
    "    'minVisibleRows': 8,\n",
    "    'sortable': True,\n",
    "    'filterable': True,\n",
    "    'highlightSelectedCell': True,\n",
    "    'highlightSelectedRow': True,\n",
    "    'show_toolbar': False\n",
    "}\n",
    "\n",
    "# Qgrid options for the BoC tables\n",
    "boc_grid_options = {\n",
    "    'fullWidthRows': True,\n",
    "    'syncColumnCellResize': True,\n",
    "    'forceFitColumns': True,\n",
    "    'defaultColumnWidth': 100,\n",
    "    'rowHeight': 23,\n",
    "    'enableColumnReorder': True,\n",
    "    'enableTextSelectionOnCells': True,\n",
    "    'editable': True,\n",
    "    'autoEdit': False,\n",
    "    'explicitInitialization': True,\n",
    "    'maxVisibleRows': 20,\n",
    "    'minVisibleRows': 8,\n",
    "    'sortable': True,\n",
    "    'filterable': True,\n",
    "    'highlightSelectedCell': True,\n",
    "    'highlightSelectedRow': True,\n",
    "    'show_toolbar': False\n",
    "}\n",
    "\n",
    "\n",
    "######################################\n",
    "# TOC CONTENTS                       #\n",
    "######################################\n",
    "\n",
    "toc_header = '''\n",
    "***\n",
    "# Table of Contents:\n",
    "'''\n",
    "toc_footer = \"***\"\n",
    "\n",
    "goback = \"[Go back to the Table of Contents](#Jovian-analysis-report)  \"\n",
    "\n",
    "qc_toc_intro = \"### Quality control metrics:\"\n",
    "multiqc_link = \"* [Quality control metrics report (MultiQC)](#quality-control-metrics-report)\"\n",
    "composition_link = \"* [Read based composition of analyzed samples](#read-based-composition-of-analyzed-samples-)\"\n",
    "\n",
    "meta_toc_intro = \"### Metagenomics\"\n",
    "krona_link = \"* [Interactive metagenomics overview](#interactive-metagenomics-overview-krona-)\"\n",
    "\n",
    "hm_toc_intro = \"#### &nbsp;&nbsp; Heatmaps\"\n",
    "hm_sup_link = \"* * [Superkingdom heatmap](#superkingdom-heatmap)\"\n",
    "hm_vir_link = \"* * [Virus heatmap](#virus-heatmaps)\"\n",
    "hm_pha_link = \"* * [Phage heatmap](#phage-heatmaps)\"\n",
    "hm_bac_link = \"* * [Bacteria heatmap](#bacteria-heatmaps)\"\n",
    "\n",
    "taxclas_toc_intro = \"### Taxonomic classification\"\n",
    "taxclas_toc_link = \"* [Classified scaffolds](#classified-scaffolds-)\"\n",
    "taxunclas_toc_link = \"* [Unclassified scaffolds](#unclassified-scaffolds)\"\n",
    "\n",
    "viro_toc_intro = \"### Viral scaffold information\"\n",
    "virhost_toc_link = \"* [Predicted virus hosts](#predicted-virus-hosts-)\"\n",
    "vir_typing_toc_intro = \"#### &nbsp;&nbsp; Viral typing results\"\n",
    "vir_typing_nov_link = \"* * [Norovirus typing tool results](#norovirus-typing-tool-output-)\"\n",
    "vir_typing_rva_link = \"* * [Rotavirus typing tool results](#rotavirus-a-typing-tool-output-)\"\n",
    "vir_typing_env_link = \"* * [Enterovirus typing tool results](#enterovirus-typing-tool-output-)\"\n",
    "vir_typing_hav_link = \"* * [Hepatitis A typing tool results](#hepatitis-a-typing-tool-output-)\"\n",
    "vir_typing_hev_link = \"* * [Hepatitis E typing tool results](#hepatitis-e-typing-tool-output-)\"\n",
    "vir_typing_hpv_link = \"* * [Human Papillomavirus typing tool results](#human-papillomavirus-typing-tool-output-)\"\n",
    "vir_typing_fla_link = \"* * [Flavirus typing tool results](#flavivirus-typing-tool-output-)\"\n",
    "\n",
    "scf_viewer_link = \"* [Interactive Scaffold viewer](#scaffold-viewer-)\"\n",
    "\n",
    "snp_table_link = \"* [Minority variant table (SNPs)](#minority-variant-table-)\"\n",
    "\n",
    "\n",
    "######################################\n",
    "# ACTUAL CONTENT                     #\n",
    "######################################\n",
    "\n",
    "mqc_cell = '''\n",
    "___\n",
    "## Quality control metrics report  \n",
    "[Open MultiQC graph in separate tab by clicking here](results/multiqc.html)  \n",
    "'''\n",
    "\n",
    "composition_cell = '''\n",
    "___\n",
    "### Read-based composition of analyzed samples:  \n",
    "[Open the barchart in separate tab by clicking here](results/Sample_composition_graph.html)  \n",
    "<br>\n",
    "**Low-quality** reads are those that did not meet the stringency settings as specified in the config file.  \n",
    "**Unclassified** reads are those that could not be assigned to a taxa.  \n",
    "**Remaining** reads are those that could not be assembled into contigs longer than the user specified minimum contig length. \n",
    "'''\n",
    "\n",
    "krona_cell = '''\n",
    "___\n",
    "## Metagenomics:\n",
    "___\n",
    "<br>\n",
    "<br>\n",
    "\n",
    "### Interactive metagenomics overview (Krona):\n",
    "[Open Krona graph in separate tab by clicking here](results/krona.html)  \n",
    "'''\n",
    "\n",
    "heatmap_cell_header = '''\n",
    "### Heatmaps:\n",
    "'''\n",
    "\n",
    "heatmap_cell_sup = '''\n",
    "#### Superkingdom heatmap  \n",
    "Open superkingdoms heatmap in a separate tab by clicking [here](results/heatmaps/Superkingdoms_heatmap.html).  \n",
    "'''\n",
    "\n",
    "heatmap_cell_vir = '''\n",
    "#### Virus heatmaps\n",
    "Open virus heatmap in a separate tab by clicking [here](results/heatmaps/Virus_heatmap.html).  \n",
    "<br>\n",
    "**Please note, many viruses have no \"`order`\" taxonomic rank so always check the \"`family`\" taxonomic rank.**\n",
    "'''\n",
    "\n",
    "heatmap_cell_pha = '''\n",
    "#### Phage heatmaps\n",
    "Open phage heatmap in a separate tab by clicking [here](results/heatmaps/Phage_heatmap.html).  \n",
    "<br>\n",
    "**Please note, many viruses have no \"`order`\" taxonomic rank so always check the \"`family`\" taxonomic rank.**\n",
    "'''\n",
    "\n",
    "heatmap_cell_bac = '''\n",
    "#### Bacteria heatmaps\n",
    "Open bacteria heatmap in a separate tab by clicking [here](results/heatmaps/Bacteria_heatmap.html).  \n",
    "'''\n",
    "\n",
    "clas_scf_cell = '''\n",
    "___\n",
    "### Classified scaffolds:\n",
    "'''\n",
    "\n",
    "unclas_scf_cell = '''\n",
    "___\n",
    "### Unclassified scaffolds (\"Dark Matter\"):\n",
    "'''\n",
    "\n",
    "virhosts_cell = '''\n",
    "___\n",
    "## Predicted virus hosts:\n",
    "___\n",
    "'''\n",
    "\n",
    "virtyping_cell = '''\n",
    "___\n",
    "## Virus typing results:\n",
    "___\n",
    "<br>\n",
    "The virus typing outputs are currently not automatically generated via Jovian due to overloading and crashing the web-service. A long-term solution is being worked on, as a work-around we've included scripts to generate these data separately. However, we kindly ask you to <u>**use this sparingly**</u> as to not overload and break the web-service. Instructions on how to generate these data can be found on the [website](https://github.com/DennisSchmitz/Jovian), [here](https://github.com/DennisSchmitz/Jovian/issues/29) and [here](https://github.com/DennisSchmitz/Jovian/issues/51).\n",
    "'''\n",
    "\n",
    "virtyping_nov_cell = '''\n",
    "### Norovirus typing tool output:  \n",
    "[Link to the norovirus typing tool](https://www.rivm.nl/mpf/typingtool/norovirus/)  \n",
    "'''\n",
    "\n",
    "virtyping_rva_cell = '''\n",
    "### Rotavirus A typing tool output:  \n",
    "[Link to the Rotavirus A typing tool](https://www.rivm.nl/mpf/typingtool/rotavirusa/)  \n",
    "'''\n",
    "\n",
    "virtyping_env_cell = '''\n",
    "### Enterovirus typing tool output:  \n",
    "[Link to the enterovirus typing tool](https://www.rivm.nl/mpf/typingtool/enterovirus/)  \n",
    "'''\n",
    "\n",
    "virtyping_hav_cell = '''\n",
    "### Hepatitis A typing tool output:  \n",
    "[Link to the hepatatis A typing tool](https://www.rivm.nl/mpf/typingtool/hav/)  \n",
    "'''\n",
    "\n",
    "virtyping_hev_cell = '''\n",
    "### Hepatitis E typing tool output:  \n",
    "[Link to the hepatatis E typing tool](https://www.rivm.nl/mpf/typingtool/hev/)  \n",
    "'''\n",
    "\n",
    "virtyping_hpv_cell = '''\n",
    "### Human Papillomavirus typing tool output:  \n",
    "[Link to the HPV typing tool](https://www.rivm.nl/mpf/typingtool/papillomavirus/)  \n",
    "'''\n",
    "\n",
    "virtyping_fla_cell = '''\n",
    "### Flavivirus typing tool output:  \n",
    "[Link to the flavivirus typing tool](https://www.rivm.nl/mpf/typingtool/flavivirus/)  \n",
    "'''\n",
    "\n",
    "boc_header_cell = '''\n",
    "___\n",
    "## Breadth of Coverage reports:\n",
    "___\n",
    "'''\n",
    "\n",
    "\n",
    "scf_viewer_cell  = '''\n",
    "___\n",
    "## Scaffold viewer:\n",
    "**Containing: SNPs and minority variants (quasispecies), predicted ORFs, depth of coverage graph, GC contents graph**\n",
    "___\n",
    "N.B. Depending on the depth of coverage of the selected contig it can be <b>(very) slow, or downright crash your browser</b>. This is a <b>client-sided</b> problem, meaning, your computer isn't powerful enough.  \n",
    "Open the scaffold viewer in a separate tab by clicking [here](results/IGVjs_index.html).  \n",
    "'''\n",
    "\n",
    "snp_table_cell = '''\n",
    "___\n",
    "## Minority variant table:\n",
    "___\n",
    "'''\n",
    "\n",
    "snp_table_empty_cell = '''\n",
    "Either no SNP's were classified, maybe because you've set the minimum allele-frequency too high? Or something went wrong, please doublecheck the logfiles below:\n",
    "    **logs/SNP_calling_[sample_name].log**\n",
    "    **logs/Concat_filtered_SNPs.log**\n",
    "'''\n",
    "\n",
    "logging_header_cell = '''\n",
    "___\n",
    "# Logging and audit-trail: \n",
    "___\n",
    "'''\n",
    "\n",
    "snakemake_report_cell = '''\n",
    "### Snakemake summary statistics\n",
    "[Open Snakemake summary statistics in separate tab by clicking here](snakemake_report.html#stats)\n",
    "'''\n",
    "\n",
    "logfile_index = '''\n",
    "<script>\n",
    "function goBack() {\n",
    "    window.history.back()\n",
    "}\n",
    "</script>\n",
    "\n",
    "<button onclick=\"goBack()\">Click this button to go back</button>\n",
    "\n",
    "<div style=\"text-align: center\">\n",
    "    <iframe src=\"results/logfiles_index.html\" width=100% height=980></iframe>\n",
    "</div>\n",
    "'''\n",
    "\n",
    "acknowledgements_header_cell = '''\n",
    "___\n",
    "# Acknowledgements:\n",
    "___\n",
    "'''\n",
    "\n",
    "######################################\n",
    "# WRITE-FUNCTIONS                    #\n",
    "######################################\n",
    "\n",
    "def toc():\n",
    "    dp(md(toc_header))\n",
    "    if os.path.exists(\"results/multiqc.html\"):\n",
    "        dp(md(qc_toc_intro))\n",
    "        dp(md(multiqc_link))\n",
    "    if os.path.exists(\"results/Sample_composition_graph.html\"):\n",
    "        dp(md(composition_link))\n",
    "    if os.path.exists(\"results/krona.html\"):\n",
    "        dp(md(meta_toc_intro))\n",
    "        dp(md(krona_link))\n",
    "    if os.path.exists(\"results/heatmaps\"):\n",
    "        dp(md(hm_toc_intro))\n",
    "        if os.path.exists(\"results/heatmaps/Superkingdoms_heatmap.html\"):\n",
    "            dp(md(hm_sup_link))\n",
    "        if os.path.exists(\"results/heatmaps/Virus_heatmap.html\"):\n",
    "            dp(md(hm_vir_link))\n",
    "        if os.path.exists(\"results/heatmaps/Phage_heatmap.html\"):\n",
    "            dp(md(hm_pha_link))\n",
    "        if os.path.exists(\"results/heatmaps/Bacteria_heatmap.html\"):\n",
    "            dp(md(hm_bac_link))\n",
    "    if os.path.exists(\"results/all_taxClassified.tsv\"):\n",
    "        dp(md(taxclas_toc_intro))\n",
    "        dp(md(taxclas_toc_link))\n",
    "        dp(md(taxunclas_toc_link))\n",
    "    if os.path.exists(\"results/all_virusHost.tsv\") or os.path.exists(\"results/typingtools\"):\n",
    "        dp(md(viro_toc_intro))\n",
    "        if os.path.exists(\"results/all_virusHost.tsv\"):\n",
    "            dp(md(virhost_toc_link))\n",
    "        if os.path.exists(\"results/typingtools\"):\n",
    "            dp(md(vir_typing_toc_intro))\n",
    "            if os.path.exists(\"results/typingtools/all_nov-TT.csv\") and os.path.getsize(\"results/typingtools/all_nov-TT.csv\") > 0:\n",
    "                dp(md(vir_typing_nov_link))\n",
    "            if os.path.exists(\"results/typingtools/all_rva-TT.csv\") and os.path.getsize(\"results/typingtools/all_rva-TT.csv\") > 0:\n",
    "                dp(md(vir_typing_rva_link))\n",
    "            if os.path.exists(\"results/typingtools/all_ev-TT.csv\") and os.path.getsize(\"results/typingtools/all_ev-TT.csv\") > 0:\n",
    "                dp(md(vir_typing_env_link))\n",
    "            if os.path.exists(\"results/typingtools/all_hav-TT.csv\") and os.path.getsize(\"results/typingtools/all_hav-TT.csv\") > 0:\n",
    "                dp(md(vir_typing_hav_link))\n",
    "            if os.path.exists(\"results/typingtools/all_hev-TT.csv\") and os.path.getsize(\"results/typingtools/all_hev-TT.csv\") > 0:\n",
    "                dp(md(vir_typing_hev_link))\n",
    "            if os.path.exists(\"results/typingtools/all_pv-TT.csv\") and os.path.getsize(\"results/typingtools/all_pv-TT.csv\") > 0:\n",
    "                dp(md(vir_typing_hpv_link))\n",
    "            if os.path.exists(\"results/typingtools/all_flavi-TT.csv\") and os.path.getsize(\"results/typingtools/all_flavi-TT.csv\") > 0:\n",
    "                dp(md(vir_typing_fla_link))\n",
    "    if os.path.exists(\"results/igv.html\"):\n",
    "        dp(md(scf_viewer_link))\n",
    "    if os.path.exists(\"results/all_filtered_SNPs.tsv\") and os.path.getsize(\"results/all_filtered_SNPs.tsv\") > 0:\n",
    "        dp(md(snp_table_link))\n",
    "    dp(md(toc_footer))\n",
    "\n",
    "\n",
    "    \n",
    "    \n",
    "\n",
    "def multiqc():\n",
    "    if os.path.exists(\"results/multiqc.html\"):\n",
    "        dp(md(mqc_cell))\n",
    "        dp(fr('results/multiqc.html', '100%', '980px'))\n",
    "        dp(md(goback))\n",
    "        \n",
    "def composition():\n",
    "    if os.path.exists(\"results/Sample_composition_graph.html\"):\n",
    "        dp(md(composition_cell))\n",
    "        dp(fr('results/Sample_composition_graph.html', '100%', '980px'))\n",
    "        dp(md(goback))\n",
    "\n",
    "def krona():\n",
    "    if os.path.exists(\"results/krona.html\"):\n",
    "        dp(md(krona_cell))\n",
    "        dp(fr('results/krona.html', '100%', '980px'))\n",
    "        dp(md(goback))\n",
    "\n",
    "def heatmaps():\n",
    "    if os.path.exists(\"results/heatmaps\"):\n",
    "        dp(md(heatmap_cell_header))\n",
    "        if os.path.exists(\"results/heatmaps/Superkingdoms_heatmap.html\"):\n",
    "            dp(md(heatmap_cell_sup))\n",
    "            dp(fr('results/heatmaps/Superkingdoms_heatmap.html', '100%', '700px'))\n",
    "            dp(md(goback))\n",
    "        if os.path.exists(\"results/heatmaps/Virus_heatmap.html\"):\n",
    "            dp(md(heatmap_cell_vir))\n",
    "            dp(fr('results/heatmaps/Virus_heatmap.html', '100%', '700px'))\n",
    "            dp(md(goback))\n",
    "        if os.path.exists(\"results/heatmaps/Phage_heatmap.html\"):\n",
    "            dp(md(heatmap_cell_pha))\n",
    "            dp(fr('results/heatmaps/Phage_heatmap.html', '100%', '700px'))\n",
    "            dp(md(goback))\n",
    "        if os.path.exists(\"results/heatmaps/Bacteria_heatmap.html\"):\n",
    "            dp(md(heatmap_cell_bac))\n",
    "            dp(fr('results/heatmaps/Bacteria_heatmap.html', '100%', '700px'))\n",
    "            dp(md(goback))\n",
    "\n",
    "def scaffoldviewer():\n",
    "    if os.path.exists(\"results/igv.html\"):\n",
    "        dp(md(scf_viewer_cell))\n",
    "        dp(fr('results/igv.html', '100%', '980px'))\n",
    "        dp(md(goback))\n",
    "\n",
    "\n",
    "def audit():\n",
    "    dp(md(logging_header_cell))\n",
    "    dp(md(\"### Sample sheet\"))\n",
    "    sheet = open(\"sample_sheet.yaml\", \"r\")\n",
    "    print(sheet.read())\n",
    "    dp(md(snakemake_report_cell))\n",
    "    if os.path.exists(\"results/snakemake_report.html\"):\n",
    "        dp(fr('results/snakemake_report.html', '100%', '980px'))\n",
    "    else:\n",
    "        print(\"\"\"\n",
    "        We couldn't find a valid snakemake report.\n",
    "        This indicates that something went wrong during the used workflow. Please run your analysis again.\n",
    "        \"\"\")\n",
    "    dp(md(\"### All log-files:\"))\n",
    "    dp(ht(logfile_index))\n",
    "    dp(md(\"### Full software list in \\\"Jovian_master\\\" environment:\"))\n",
    "    conda_log = open(\"results/log_conda.txt\", \"r\")\n",
    "    print(conda_log.read())\n",
    "    dp(md(\"### Database versions:\"))\n",
    "    db_log = open(\"results/log_db.txt\", \"r\")\n",
    "    print(db_log.read())\n",
    "    dp(md(\"### Unique methodological \\\"fingerprint\\\":\"))\n",
    "    id_log = open(\"results/log_git.txt\", \"r\")\n",
    "    print(id_log.read())\n",
    "    dp(md(\"### Snakemake config files:\"))\n",
    "    conf_log = open(\"results/log_config.txt\", \"r\")\n",
    "    print(conf_log.read())\n",
    "    dp(md(goback))\n",
    "\n",
    "\n",
    "def acks():\n",
    "    dp(md(acknowledgements_header_cell))\n",
    "    with open(\"files/acknowledgements.md\", \"r\") as ack_base:\n",
    "        ack_list = ack_base.read()\n",
    "    dp(md(ack_list))\n",
    "    with open(\"files/authors.md\", \"r\") as ath_base:\n",
    "        ath_list = ath_base.read()\n",
    "    dp(md(ath_list))\n",
    "\n",
    "\n",
    "### interactive tables and/or grids            \n",
    "class Classified_grid:\n",
    "    def dataframe(clas):\n",
    "        if os.path.exists(\"results/all_taxClassified.tsv\"):\n",
    "            clas.df = pd.read_csv('results/all_taxClassified.tsv', sep = \"\\t\")\n",
    "            clas.grid = qgrid.show_grid(clas.df, grid_options=grid_options)\n",
    "            return\n",
    "    def draw(clas):\n",
    "        clas.dataframe()\n",
    "        if os.path.exists(\"results/all_taxClassified.tsv\"):\n",
    "            dp(md(clas_scf_cell))\n",
    "            dp(clas.grid)\n",
    "            dp(md(goback))\n",
    "\n",
    "class Unclassified_grid:\n",
    "    def dataframe(unclas):\n",
    "        if os.path.exists(\"results/all_taxUnclassified.tsv\"):\n",
    "            unclas.df = pd.read_csv('results/all_taxUnclassified.tsv', sep = \"\\t\")\n",
    "            unclas.grid = qgrid.show_grid(unclas.df, grid_options=grid_options)\n",
    "            return\n",
    "    def draw(unclas):\n",
    "        unclas.dataframe()\n",
    "        if os.path.exists(\"results/all_taxClassified.tsv\"):\n",
    "            dp(md(unclas_scf_cell))\n",
    "            dp(unclas.grid)\n",
    "            dp(md(goback))\n",
    "\n",
    "class Predict_virhosts:\n",
    "    def dataframe(predicts):\n",
    "        if os.path.exists(\"results/all_virusHost.tsv\"):\n",
    "            predicts.df = pd.read_csv('results/all_virusHost.tsv', sep = \"\\t\")\n",
    "            predicts.grid = qgrid.show_grid(predicts.df, grid_options=grid_options)\n",
    "            return\n",
    "    def draw(predicts):\n",
    "        predicts.dataframe()\n",
    "        if os.path.exists(\"results/all_virusHost.tsv\"):\n",
    "            dp(md(virhosts_cell))\n",
    "            dp(predicts.grid)\n",
    "            dp(md(goback))\n",
    "\n",
    "def virtypingheader():\n",
    "    if os.path.exists(\"results/typingtools\"):\n",
    "        dp(md(virtyping_cell))\n",
    "\n",
    "\n",
    "class nov_typing:\n",
    "    def dataframe(nov):\n",
    "        if os.path.exists(\"results/typingtools/all_nov-TT.csv\") and os.path.getsize(\"results/typingtools/all_nov-TT.csv\") > 0:\n",
    "            nov.df = pd.read_csv('results/typingtools/all_nov-TT.csv')\n",
    "            nov.grid = qgrid.show_grid(nov.df, grid_options=grid_options)\n",
    "            return\n",
    "    def draw(nov):\n",
    "        nov.dataframe()\n",
    "        if os.path.exists(\"results/typingtools/all_nov-TT.csv\") and os.path.getsize(\"results/typingtools/all_nov-TT.csv\") > 0:\n",
    "            dp(md(virtyping_nov_cell))\n",
    "            dp(nov.grid)\n",
    "            dp(md(goback))\n",
    "\n",
    "class rva_typing:\n",
    "    def dataframe(rva):\n",
    "        if os.path.exists(\"results/typingtools/all_rva-TT.csv\") and os.path.getsize(\"results/typingtools/all_rva-TT.csv\") > 0:\n",
    "            rva.df = pd.read_csv('results/typingtools/all_rva-TT.csv')\n",
    "            rva.grid = qgrid.show_grid(rva.df, grid_options=grid_options)\n",
    "            return\n",
    "    def draw(rva):\n",
    "        rva.dataframe()\n",
    "        if os.path.exists(\"results/typingtools/all_rva-TT.csv\") and os.path.getsize(\"results/typingtools/all_rva-TT.csv\") > 0:\n",
    "            dp(md(virtyping_rva_cell))\n",
    "            dp(rva.grid)\n",
    "            dp(md(goback))\n",
    "\n",
    "class env_typing:\n",
    "    def dataframe(env):\n",
    "        if os.path.exists(\"results/typingtools/all_ev-TT.csv\") and os.path.getsize(\"results/typingtools/all_ev-TT.csv\") > 0:\n",
    "            env.df = pd.read_csv('results/typingtools/all_ev-TT.csv')\n",
    "            env.grid = qgrid.show_grid(env.df, grid_options=grid_options)\n",
    "            return\n",
    "    def draw(env):\n",
    "        env.dataframe()\n",
    "        if os.path.exists(\"results/typingtools/all_ev-TT.csv\") and os.path.getsize(\"results/typingtools/all_ev-TT.csv\") > 0:\n",
    "            dp(md(virtyping_env_cell))\n",
    "            dp(env.grid)\n",
    "            dp(md(goback))\n",
    "\n",
    "class hav_typing:\n",
    "    def dataframe(hav):\n",
    "        if os.path.exists(\"results/typingtools/all_hav-TT.csv\") and os.path.getsize(\"results/typingtools/all_hav-TT.csv\") > 0:\n",
    "            hav.df = pd.read_csv('results/typingtools/all_hav-TT.csv')\n",
    "            hav.grid = qgrid.show_grid(hav.df, grid_options=grid_options)\n",
    "            return\n",
    "    def draw(hav):\n",
    "        hav.dataframe()\n",
    "        if os.path.exists(\"results/typingtools/all_hav-TT.csv\") and os.path.getsize(\"results/typingtools/all_hav-TT.csv\") > 0:\n",
    "            dp(md(virtyping_hav_cell))\n",
    "            dp(hav.grid)\n",
    "            dp(md(goback))\n",
    "\n",
    "class hev_typing:\n",
    "    def dataframe(hev):\n",
    "        if os.path.exists(\"results/typingtools/all_hev-TT.csv\") and os.path.getsize(\"results/typingtools/all_hev-TT.csv\") > 0:\n",
    "            hev.df = pd.read_csv('results/typingtools/all_hev-TT.csv')\n",
    "            hev.grid = qgrid.show_grid(hev.df, grid_options=grid_options)\n",
    "            return\n",
    "    def draw(hev):\n",
    "        hev.dataframe()\n",
    "        if os.path.exists(\"results/typingtools/all_hev-TT.csv\") and os.path.getsize(\"results/typingtools/all_hev-TT.csv\") > 0:\n",
    "            dp(md(virtyping_hev_cell))\n",
    "            dp(hev.grid)\n",
    "            dp(md(goback))\n",
    "\n",
    "class hpv_typing:\n",
    "    def dataframe(hpv):\n",
    "        if os.path.exists(\"results/typingtools/all_pv-TT.csv\") and os.path.getsize(\"results/typingtools/all_pv-TT.csv\") > 0:\n",
    "            hpv.df = pd.read_csv('results/typingtools/all_pv-TT.csv')\n",
    "            hpv.grid = qgrid.show_grid(hpv.df, grid_options=grid_options)\n",
    "            return\n",
    "    def draw(hpv):\n",
    "        hpv.dataframe()\n",
    "        if os.path.exists(\"results/typingtools/all_pv-TT.csv\") and os.path.getsize(\"results/typingtools/all_pv-TT.csv\") > 0:\n",
    "            dp(md(virtyping_hpv_cell))\n",
    "            dp(hpv.grid)\n",
    "            dp(md(goback))\n",
    "\n",
    "class fla_typing:\n",
    "    def dataframe(fla):\n",
    "        if os.path.exists(\"results/typingtools/all_flavi-TT.csv\") and os.path.getsize(\"results/typingtools/all_flavi-TT.csv\") > 0:\n",
    "            fla.df = pd.read_csv('results/typingtools/all_flavi-TT.csv')\n",
    "            fla.grid = qgrid.show_grid(fla.df, grid_options=grid_options)\n",
    "            return\n",
    "    def draw(fla):\n",
    "        fla.dataframe()\n",
    "        if os.path.exists(\"results/typingtools/all_flavi-TT.csv\") and os.path.getsize(\"results/typingtools/all_flavi-TT.csv\") > 0:\n",
    "            dp(md(virtyping_fla_cell))\n",
    "            dp(fla.grid)\n",
    "            dp(md(goback))\n",
    "\n",
    "class snp_variants:\n",
    "    def dataframe(snps):\n",
    "        if os.path.exists(\"results/all_filtered_SNPs.tsv\") and os.path.getsize(\"results/all_filtered_SNPs.tsv\") > 0:\n",
    "            snps.df = pd.read_csv('results/all_filtered_SNPs.tsv', sep=\"\\t\")\n",
    "            snps.grid = qgrid.show_grid(snps.df, grid_options=grid_options)\n",
    "            return\n",
    "    def draw(snps):\n",
    "        snps.dataframe()\n",
    "        if os.path.exists(\"results/all_filtered_SNPs.tsv\") and os.path.getsize(\"results/all_filtered_SNPs.tsv\") > 0:\n",
    "            dp(md(snp_table_cell))\n",
    "            dp(snps.grid)\n",
    "            dp(md(goback))\n",
    "        else:\n",
    "            dp(md(snp_table_cell))\n",
    "            dp(md(snp_table_empty_cell))\n",
    "\n",
    "def bocheaders():\n",
    "    if os.path.exists(\"results/BoC_int.tsv\") or os.path.exists(\"results/BoC_pct.tsv\"):\n",
    "        dp(md(boc_header_cell))\n",
    "            \n",
    "class breadth_of_cov_ints:\n",
    "    def dataframe(ints):\n",
    "        if os.path.exists(\"results/BoC_int.tsv\") and os.path.getsize(\"results/BoC_int.tsv\") > 0:\n",
    "            ints.df = pd.read_csv('results/BoC_int.tsv', sep=\"\\t\")\n",
    "            ints.df.set_index('Sample_name', inplace=True)\n",
    "            ints.grid = qgrid.show_grid(ints.df, grid_options=boc_grid_options)\n",
    "            return\n",
    "    def draw(ints):\n",
    "        ints.dataframe()\n",
    "        if os.path.exists(\"results/BoC_int.tsv\") and os.path.getsize(\"results/BoC_int.tsv\") > 0:\n",
    "            dp(md(\"### Breadth of Coverage overview in nucleotide-numbers\"))\n",
    "            dp(ints.grid)\n",
    "            dp(md(goback))\n",
    "\n",
    "class breadth_of_cov_pcts:\n",
    "    def dataframe(pcts):\n",
    "        if os.path.exists(\"results/BoC_pct.tsv\") and os.path.getsize(\"results/BoC_pct.tsv\") > 0:\n",
    "            pcts.df = pd.read_csv('results/BoC_pct.tsv',  sep=\"\\t\")\n",
    "            pcts.df.set_index('Sample_name', inplace=True)\n",
    "            pcts.grid = qgrid.show_grid(pcts.df, grid_options=boc_grid_options)\n",
    "            return\n",
    "    def draw(pcts):\n",
    "        pcts.dataframe()\n",
    "        if os.path.exists(\"results/BoC_pct.tsv\") and os.path.getsize(\"results/BoC_pct.tsv\") > 0:\n",
    "            dp(md(\"### Breath of Coverage overview in percentages\"))\n",
    "            dp(pcts.grid)\n",
    "            dp(md(goback))\n",
    "\n",
    "classified_scaffolds = Classified_grid()\n",
    "unclassified_scaffolds = Unclassified_grid()\n",
    "predicted_virushosts = Predict_virhosts()\n",
    "virus_typing_nov = nov_typing()\n",
    "virus_typing_rva = rva_typing()\n",
    "virus_typing_env = env_typing()\n",
    "virus_typing_hav = hav_typing()\n",
    "virus_typing_hev = hev_typing()\n",
    "virus_typing_hpv = hpv_typing()\n",
    "virus_typing_fla = fla_typing()\n",
    "minority_variant_table = snp_variants()\n",
    "boc_table_ints = breadth_of_cov_ints()\n",
    "boc_table_pcts = breadth_of_cov_pcts()\n",
    "\n",
    "### actually draw the contents\n",
    "\n",
    "#toc()\n",
    "multiqc()\n",
    "composition()\n",
    "krona()\n",
    "heatmaps()\n",
    "\n",
    "classified_scaffolds.draw()\n",
    "unclassified_scaffolds.draw()\n",
    "\n",
    "predicted_virushosts.draw()\n",
    "\n",
    "virtypingheader()\n",
    "virus_typing_nov.draw()\n",
    "virus_typing_rva.draw()\n",
    "virus_typing_env.draw()\n",
    "virus_typing_hav.draw()\n",
    "virus_typing_hev.draw()\n",
    "virus_typing_hpv.draw()\n",
    "virus_typing_fla.draw()\n",
    "\n",
    "bocheaders()\n",
    "boc_table_ints.draw()\n",
    "boc_table_pcts.draw()\n",
    "\n",
    "\n",
    "scaffoldviewer()\n",
    "\n",
    "minority_variant_table.draw()\n",
    "\n",
    "audit()\n",
    "acks()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "___\n",
    "Jovian is available on [GitHub](https://github.com/DennisSchmitz/Jovian) under a [AGPL license](https://www.gnu.org/licenses/agpl-3.0). The virus-typing tools are public services hosted by the [RIVM](https://www.rivm.nl/en) and developed independently of Jovian.\n",
    "___"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "deleteable": false,
    "init_cell": true,
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "dp(md(goback))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "deleteable": false,
    "init_cell": true,
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "%%javascript\n",
    "$('<div id=\"toc\"></div>').css({position: 'fixed', top: '120px', left: 0}).appendTo(document.body);\n",
    "$.getScript('https://kmahelona.github.io/ipython_notebook_goodies/ipython_notebook_toc.js');"
   ]
  }
 ],
 "metadata": {
  "hide_input": true,
  "kernelspec": {
   "display_name": "Jovian",
   "language": "python",
   "name": "jovian_master"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.7"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": false,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {
    "height": "calc(100% - 180px)",
    "left": "10px",
    "top": "150px",
    "width": "210px"
   },
   "toc_section_display": true,
   "toc_window_display": true
  },
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "position": {
    "height": "413px",
    "left": "1476px",
    "right": "20px",
    "top": "120px",
    "width": "350px"
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
